libyul/AsmAnalysis.cpp

```diff
@@ -293,11 +293,13 @@ vector<YulString> AsmAnalyzer::operator()(FunctionCall const& _funCall)
 		);
 
 	vector<YulString> argTypes;
-	for (auto const& arg: _funCall.arguments | boost::adaptors::reversed)
+	for (size_t i = _funCall.arguments.size(); i > 0; i--)
```

Comment:
Can we create a helper for that? I think the `i - 1` is really dangerous. Is there something like python's enumerate? Then we could use `for (auto&& [i, arg]: enumerate(_funCall.arguments) | reversed)`

Reply:
Yes absolutely. I wasn't happy with that either.

Reply:
http://reedbeta.com/blog/python-like-enumerate-in-cpp17/

Reply:
@christianparpart do you have a good idea how this could be done?

Reply:
Maybe just
```
for (size_t i : boost::counting_range(0, _funCall.arguments.size()) | boost::adaptors::reversed)
{
  auto&& arg = _funCall.arguments[i];
```

I could look into helpers that give you tuples of index and value that are robust against reversing - or even helpers that let you iterate two vectors simultanously, so you could use ``auto&& [needsLiteralArgument, arg]: someWrapper(*needsLiteralArguments, _funCall.arguments)`` - that's possible, but it can get quite messy quickly :-).

Reply:
Double iteration is nice, but `needsLiteralArguments` can be nullptr

Reply:
Github kind of lacked in displaying the comments for me again...

> http://reedbeta.com/blog/python-like-enumerate-in-cpp17/

Yeah, stuff like that works - but those e.g. won't be reversible, for that you need to add reverse iterators as well, etc. - that's what I meant by messy :-). But might not be too bad.

Reply:
Did you see the link I posted above?

Reply:
By the way: the solution in the link always creates a copy - a more sophisticated solution would use references for lvalues and move for rvalues (again part of the messy-ness ;-))

Reply:
I think it's fine as is. `i - 1` is only used twice and the `for` line is very clear

Reply:
Maybe change `i` to something else though? `i` is also used below, might look confusing visually

Reply:
We have tons of other places in the code where an `enumerate` or "iterate two lists at the same time" would be very handy.